app/src/play/java/de/danoeh/antennapod/cast/CustomMRControllerDialog.java

```diff
@@ -0,0 +1,347 @@
+            }
+        });
+
+        viewsCreated = true;
+        updateViews();
+        return rootView;
+    }
+
+    private void updateViews() {
+        if (!viewsCreated || token == null || mediaController == null) {
+            rootView.setVisibility(View.GONE);
+            return;
+        }
+        MediaMetadataCompat metadata = mediaController.getMetadata();
+        MediaDescriptionCompat description = metadata == null ? null : metadata.getDescription();
+        if (description == null) {
+            rootView.setVisibility(View.GONE);
+            return;
+        }
+
+        PlaybackStateCompat state = mediaController.getPlaybackState();
+        MediaRouter.RouteInfo route = MediaRouter.getInstance(getContext()).getSelectedRoute();
+
+        CharSequence title = description.getTitle();
+        boolean hasTitle = !TextUtils.isEmpty(title);
+        CharSequence subtitle = description.getSubtitle();
+        boolean hasSubtitle = !TextUtils.isEmpty(subtitle);
+
+        boolean showTitle = false;
+        boolean showSubtitle = false;
+        if (route.getPresentationDisplayId() != MediaRouter.RouteInfo.PRESENTATION_DISPLAY_ID_NONE) {
+            // The user is currently casting screen.
+            titleView.setText(android.support.v7.mediarouter.R.string.mr_controller_casting_screen);
+            showTitle = true;
+        } else if (state == null || state.getState() == PlaybackStateCompat.STATE_NONE) {
+            // Show "No media selected" as we don't yet know the playback state.
+            // (Only exception is bluetooth where we don't show anything.)
+            if (!route.isDeviceTypeBluetooth()) {
+                titleView.setText(android.support.v7.mediarouter.R.string.mr_controller_no_media_selected);
+                showTitle = true;
+            }
+        } else if (!hasTitle && !hasSubtitle) {
+            titleView.setText(android.support.v7.mediarouter.R.string.mr_controller_no_info_available);
+            showTitle = true;
+        } else {
+            if (hasTitle) {
+                titleView.setText(title);
+                showTitle = true;
+            }
+            if (hasSubtitle) {
+                subtitleView.setText(subtitle);
+                showSubtitle = true;
+            }
+        }
+        if (showSubtitle) {
+            titleView.setSingleLine();
+        } else {
+            titleView.setMaxLines(2);
+        }
+        titleView.setVisibility(showTitle ? View.VISIBLE : View.GONE);
+        subtitleView.setVisibility(showSubtitle ? View.VISIBLE : View.GONE);
+
+        updateState();
+
+        if(rootView.getVisibility() != View.VISIBLE) {
+            artView.setVisibility(View.GONE);
+            rootView.setVisibility(View.VISIBLE);
+        }
+
+        if (fetchArtTask != null) {
+            fetchArtTask.cancel(true);
+        }
+
+        fetchArtTask = new FetchArtTask(description);
+        fetchArtTask.execute();
+    }
+
+    private void updateState() {
```

Comment:
maybe this is just NPE paranoia, but would it be possible that ``updateState`` is called before ``onCreateMediaControlView`` via ``MediaControllerCompat.Callback``?